### ISSUE TITLE
Support Advanced Hovers

### ISSUE THREAD

**Issue by JustinGrote** [PowerShell/vscode-powershell#3026]
**Summary of the new feature**

The existing powershell hover for commands/functions is very basic and only provides the synopsis. VScode now supports colorized advanced hovers, and Omnisharp already implements this as a reference.
https://code.visualstudio.com/api/language-extensions/programmatic-language-features#show-hovers
![image](https://user-images.githubusercontent.com/15258962/97047318-9efca080-152d-11eb-9008-91fd91c24875.png)

Powershell should support the same for functions and methods, displaying colorized syntax info about the function, and ideally not just the synopsis but parameter descriptions. The verbosity could be controlled via a setting.

**Proposed technical implementation details (optional)**

This should be done after the parser rewrite. It should probably be tagged a good-first-issue as well, I'd be willing to give it a shot once the parser rewrite is done to avoid a moving target.

**Comment by MartinGC94** [comment]
It would also be nice if hovering over a variable/property would show the type like Visual Studio does. If the variable/property doesn't currently have a value the internal PowerShell type inference code could be used.

**Comment by JustinGrote** [comment]
@MartinGC94 that feature would be better suited as an inlay hint but I agree, we could definitely offer a high confidence and best effort inference here, it's one of my to-dos after inline breakpointing.

https://www.youtube.com/watch?v=uvrIFZYW7eg

**Comment by MartinGC94** [comment]
I hadn't heard about inlay hints before. They seem neat, though I'd still prefer the hover tooltips but maybe I'll change my mind if I try to use those hints.
> Also I'm going to use this opportunity to MASSIVELY thank you for all of your completion and autocomplete fixups you've done over the past year or two, they make my quality of life so much better :)

Thanks, it's been a great way for me to sharpen my C# skills and I of course also love the UX improvements myself.

**Comment by JustinGrote** [comment]
@MartinGC94 they are awesome and fully implemented in Typescript/F#/C# extensions. Try opening a file and hitting ctrl-alt. You can also configure them to be persistently visible for various types.

**Comment by JustinGrote** [comment]
@MartinGC94 I have a basic prototype wired up here:
https://github.com/PowerShell/PowerShellEditorServices/tree/justingrote/feature/inlayHints
![image](https://github.com/PowerShell/vscode-powershell/assets/15258962/411e8959-bc01-49e3-86b0-fe96f7bcd85d)

Can you point me to docs or the API that is the best way to find the inferred types of variables and parameters in PowerShell? Ideally one that does a single pass of a whole document/script as efficiently as possible rather than a ton of individual AST visits.

**Comment by JustinGrote** [comment]
OK, after some experimentation and discussion with @SeeminglyScience, doing inlay hints is probably not feasible, as the current implementation would basically have to get a `CompletionResult` for pretty much every token (we could potentially do some smart scope deduplication) that is on the screen and refresh all of it either when the document changes or the view changes (if we limit completions to just what is on screen)

Until PowerShell has a performant API that can provide a visitor for gathering this kind of info, it's doable but it would be slow AF and not super useful.

Enhancing the hovers however is definitely doable and something I'll explore some more.

**Comment by andyleejordan** [comment]
I would love to see the existing hovers made better. Not sure if you saw https://github.com/PowerShell/PowerShellEditorServices/pull/2127. There's a lot we could do.

**Comment by JustinGrote** [comment]
@andyleejordan you probably sent it to me and it fell off my to-look list :)

**Comment by JustinGrote** [comment]
I was investigating this a but further today.

While we could theoretically offer this today, unfortunately to do inline hints "simply", requires running completion for every variable/parameter symbol in the document, which even if we defer in the pipeline, will result in a *lot* of compute and while fine for relatively small scripts, would end up being very slow for large scripts. We would be able to optimize by potentially only doing deltas after that, but it's still really heavyweight by virtue of that it would have to run within the runspace.

The alternative is utilize several of the additions from the rename provider to do static AST analysis. For example:

```
[int]$x

$x <-- can show int as inlay hint
```

We effectively expand on the rename `FindVariableDefinition` functions and, if there is a hard type defined, we can relatively safely display inlay hints for subsequent references. This would be another "best effort" feature, and we could still implement the inlay hint feature but maybe cut it off at only the visible document and refresh when requested.

I'll experiment with this more.

**Comment by MartinGC94** [comment]
Another place where hovers would be useful is type expressions where it could show the full typename, this would be especially useful in the future when type aliases are implemented. For example:

```
using type sList = System.Collections.Generic.List[string]

# Hover shows System.Collections.Generic.List[string]
[sList]::new()
```

but even today with `using namespace` or type accelerators it would be useful.

**Comment by JustinGrote** [comment]
The tricky thing is that right now with vscode, hovers aren't "additive", meaning we can't show a hover without gathering all the info. So to make this work without annoying a lot of people, it'll have to be opt in, and they'll need to be aware it'll be slower to pop up than others.

I haven't looked if vscode has some more streaming option natively, but it's definitely not part of the hover LSP spec.

**Comment by MartinGC94** [comment]
Will it really be that much slower though? You would only have to run the parser, find the ast the user is hovering over, then if it's a TypeExpressionAst, run `GetReflectionType`.

This:
```
$BaseAst = [System.Management.Automation.Language.Parser]::ParseFile("C:\Test.ps1", [ref] $null, [ref] $null)
($BaseAst.FindAll({param($ast) $ast -is [System.Management.Automation.Language.TypeExpressionAst]},$true) | select -Last 1).TypeName.GetReflectionType()
```
took 116ms to run on my laptop the first time and 53ms the second time. That seems fast enough for hovering because people are used to having to wait a bit for hovers.

**Comment by JustinGrote** [comment]
@MartinGC94 sorry, I wasn't speaking specifically to your suggested type updates, which I think are certainly reasonabe.
